nd4j/nd4j-backends/nd4j-tests/src/test/java/org/nd4j/linalg/factory/ops/NDBaseTest.java

```diff
@@ -0,0 +1,923 @@
+        INDArray y = base.oneHot(x, 1, 0, 1.0, 0.0);
+        INDArray y_exp = Nd4j.createFromArray(new float[][]{{1.0f, 0.0f, 0.0f}});
+        assertEquals(y, y_exp);
+
+        y = base.oneHot(x, 1);
+        y_exp = Nd4j.createFromArray(new float[][]{{1.0f},{ 0.0f}, {0.0f}});
+        assertEquals(y, y_exp);
+
+        y = base.oneHot(x, 1, 0, 1.0, 0.0, DataType.DOUBLE);
+        y_exp = Nd4j.createFromArray(new double[][]{{1.0, 0.0, 0.0}});
+        assertEquals(y, y_exp); //TODO: Looks like we're getting back the wrong datatype.
+    }
+
+    @Test
+    public void testOnesLike() {
+        NDBase base = new NDBase();
+        INDArray x = Nd4j.create(new double[]{3, 3});
+        INDArray y = base.onesLike(x);
+        INDArray  y_exp = Nd4j.createFromArray(new double[]{1.0, 1.0});
+        assertEquals(y, y_exp);
+
+        y = base.onesLike(x, DataType.LONG);
+        y_exp = Nd4j.createFromArray(new long[]{1, 1});
+        assertEquals(y, y_exp); //TODO: Getting back a double array, not a long.
+    }
+
+    @Test
+    public void testPermute() {
+        NDBase base = new NDBase();
+        INDArray x = Nd4j.create(new double[]{1, 2, 3});
+        INDArray y = base.permute(x, 2, 0, 1);
+        //TODO: fix. no output arrays were provided and calculateOutputShape failed to execute
+
+    }
+
+    @Test
+    public void testProd() {
+        NDBase base = new NDBase();
+        INDArray x = Nd4j.eye(3);
+        INDArray y = base.prod(x, 0);
+        INDArray y_exp = Nd4j.createFromArray(new float[]{0.0f, 0.0f, 0.0f});
+        assertEquals(y, y_exp);
+
+        y = base.prod(x, true, 0);
+        y_exp = Nd4j.createFromArray(new float[][]{{0.0f, 0.0f, 0.0f}});
+        assertEquals(y, y_exp);
+    }
+
+    @Test
+    public void testRange() {
+        NDBase base = new NDBase();
+        INDArray y = base.range(0.0, 3.0, 1.0, DataType.DOUBLE);
+        INDArray y_exp = Nd4j.createFromArray(new double[]{0.0, 1.0, 2.0f});
+        assertEquals(y, y_exp); //TODO: Asked for DOUBLE, got back a FLOAT Array.
+    }
+
+    @Test
+    public void testRank() {
+        NDBase base = new NDBase();
+        INDArray x = Nd4j.eye(3);
+        INDArray y = base.rank(x);
+        INDArray y_exp = Nd4j.scalar(2);
+        System.out.println(y);
+        assertEquals(y, y_exp);
+    }
+
+    @Test
+    public void testRepeat() {
+        NDBase base = new NDBase();
+        INDArray x = Nd4j.eye(3);
+        INDArray y = base.repeat(x, 0);
+        //TODO: fix, crashes the JVM.
+    }
+
+
+    @Test
+    public void testReplaceWhere() {
+        // test from BooleanIndexingTest.
+        NDBase base = new NDBase();
+        INDArray array = Nd4j.create(new double[] {1, 2, 0, 4, 5});
+        INDArray comp = Nd4j.create(new double[] {1, 2, 3, 4, 5});
+        INDArray y = base.replaceWhere(array, comp, Conditions.lessThan(1));
+        assertEquals(comp, array);
+        assertEquals(y, array); // in fact y points to the same INDArray. Is that expected?
```

Comment:
As a general rule, it shouldn't be.
assertNotSame(y, array) would pick that up in the test, FWIW